Zend/zend_compile.c

```diff
@@ -2443,6 +2448,16 @@ static void zend_emit_return_type_check(
 			return;
 		}
 
+		/* `return` is illegal in a noreturn function */
+		if (ZEND_TYPE_CONTAINS_CODE(type, IS_NEVER)) {
+			if (!implicit) {
```

Comment:
Make this a `ZEND_ASSERT(!implicit)` -- I believe you never invoke this function in that case.